core/services/directrequest/delegate.go

```diff
@@ -123,44 +161,88 @@ func (*listener) OnConnect() {}
 // OnDisconnect complies with log.Listener
 func (*listener) OnDisconnect() {}
 
-// OnConnect complies with log.Listener
-func (l *listener) HandleLog(lb log.Broadcast) {
-	was, err := lb.WasAlreadyConsumed()
-	if err != nil {
-		logger.Errorw("DirectRequestListener: could not determine if log was already consumed", "error", err)
-		return
-	} else if was {
-		return
+func (l *listener) OnNewLongestChain(ctx context.Context, head models.Head) {
+	select {
+	case l.chHeads <- head:
+	default:
 	}
+}
 
-	logJobSpecID := lb.RawLog().Topics[1]
-	if logJobSpecID == (common.Hash{}) || logJobSpecID != l.onChainJobSpecID {
-		logger.Debugw("Skipping Run for Log with wrong Job ID", "logJobSpecID", logJobSpecID, "actualJobID", l.onChainJobSpecID)
-		return
-	}
+func (l *listener) HandleLog(lb log.Broadcast) {
+	l.mbLogs.Deliver(lb)
+}
 
-	log := lb.DecodedLog()
-	if log == nil || reflect.ValueOf(log).IsNil() {
-		logger.Error("HandleLog: ignoring nil value")
-		return
+func (l *listener) run() {
+	for {
+		select {
+		case <-l.chStop:
+			return
+		case head := <-l.chHeads:
+			l.handleReceivedLogs(head)
+		}
 	}
+}
 
-	switch log := log.(type) {
-	case *oracle_wrapper.OracleOracleRequest:
-		l.handleOracleRequest(log)
-		err = lb.MarkConsumed()
-		if err != nil {
-			logger.Errorf("Error marking log as consumed: %v", err)
+func (l *listener) handleReceivedLogs(head models.Head) {
+	oldEnough := isOldEnoughConstructor(head, l.minConfirmations)
+	for {
+		i := l.mbLogs.RetrieveIf(oldEnough)
+		if i == nil {
+			return
 		}
-	case *oracle_wrapper.OracleCancelOracleRequest:
-		l.handleCancelOracleRequest(log)
-		err = lb.MarkConsumed()
+		lb, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", lb)
```

Comment:
I think it would actually be reasonable to panic here

Reply:
ok